### ISSUE TITLE
How can I convert the bounding boxes to correct values using training model ?

### ISSUE THREAD

**Issue by lazydroid** [fizyr/keras-retinanet#909]
![bounding_boxes_2019-02-19](https://user-images.githubusercontent.com/8067211/52997659-8557ab80-3464-11e9-8d1b-02d2e8fe728f.png)

I haven't converted the training model to the inference one. So the results for the bounding boxes are as above (the picture taken from #872). What is the meaning of these numbers and what is the simplest way to convert them to correct values?

I could get through the source code and try to figure out the meaning and the conversion method, but I would really appreciate to get a simple explanation that I can code and get the results without the model conversion -- please, don't tell me to convert the model, for some reason this is not possible in my case.

Thanks in advance!

**Comment by hgaiser** [comment]
I wonder why converting the model is not an option in your case, all it does is add a few layers to decode these values to boxes and to perform non-maximum suppression (which can optionally be disabled as well).

Anyway, these values represent the regression targets as estimated by the network. Taking your very first value as an example (`-0.00086247`), this represents the change in `x1` (or `x_min`) that should be applied to the anchor to go to the predicted box. Check [here](https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/backend/common.py#L45) to see how the conversion is being done. The actual value of `x1` would become:

```
x1 = x1_anchor + (-0.00086247 * 0.2 + 0.0) * anchor_width
```

This requires to know the anchor coordinates for which these regression values are applied.

The reason for the `0.2` and `0.0` is to normalize the regression targets to have 0 mean and a std of 1. Neural nets like normalized values, saves them the trouble of having to normalize it themselves.

Having said all this, I really think it is easier if you re-use the available layers, but that's up to you. I'll close this issue as there is no issue for our code here.

**Comment by lazydroid** [comment]
> This requires to know the anchor coordinates

@hgaiser Thank you very much for your help, I really appreciate your answers!

Am I correct to assume that anchor coordinates are always same for the same anchor?

I've got 57600 possible boxes as a network output, I assume there's an anchor grid evenly distributed over the image, with the fixed positions, could you please point me to the location where the anchor positions are calculated? I use fixed size 640x480 images as an input.

Regarding reusing available layers, I'm trying to use this network in an embedded environment, which supports only a few of TF layers, and while most of the layers required for the detection are in, a lot of layers used for NMS and filtering are not.

**Comment by lazydroid** [comment]
And another question would be about the classification. I've got (57600,3) classification output (I have 3 classes), however, the highest value is 0.00238 -- shall these value be scaled back somehow or does this mean there are no detections on the input image?

The same image processed through the converted model gives a few detections, so I expected the classification output to contain a few values close to 1.0 -- am I somehow wrong about that ?

**Comment by hgaiser** [comment]
> Am I correct to assume that anchor coordinates are always same for the same anchor?

Yes, they're generated using [this](https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/utils/anchors.py#L201) function.

> I've got 57600 possible boxes as a network output, I have found the place where the anchor positions are calculated, are those 43200 + 10800 + 2700 + 720 + 180 anchors in the same order as the `regression` and `classification` outputs?

Yes. They are the anchors for P3, P4, P5, P6 and P7 respectively.



> And another question would be about the classification. I've got (57600,3) classification output (I have 3 classes), however, the highest value is 0.00238 -- shall these value be scaled back somehow or does this mean there are no detections on the input image?

The classification has no scaling applied to it when converting to an inference model. Perhaps your preprocessing for your image is not correct?

**Comment by lazydroid** [comment]
@hgaiser I read my image, then change channel order to BGR, convert to `float` and subtract the mean channel value for ImageNet (same as `prepare_image`, I believe).

Am I supposed to normalize (change -127...127 range to -1...1 by dividing every pixel by 127.0 ?

**Comment by lazydroid** [comment]
@hgaiser Well, just in case someone struggles with the same problem, here's my 5 cents. Regression parameters come in a different order, so to get correct boxes, `anchors` and `regression` should be done like this (note the numbers and minus signs):
```
		x1 = anc[0] - (reg[2] * 0.2 + 0.0) * anchor_width
		y1 = anc[1] - (reg[3] * 0.2 + 0.0) * anchor_height
		x2 = anc[2] - (reg[0] * 0.2 + 0.0) * anchor_width
		y2 = anc[3] - (reg[1] * 0.2 + 0.0) * anchor_height
```

**Comment by Cavasurg** [comment]
Dear Lenik Terenin!
I greatly apologize for the off-top, but it is very difficult to contact you, although I have repeatedly tried to do it. I have a huge request for you, as the owner of the site http://www.samosud.org. Your resource contains copies of the court decisions: http://www.samosud.org/case_969293378 and http://www.samosud.org/case_849613734, in which I appear as the complainant in the labor dispute. The problem is that while somebody enters my name into the Yandex search box, the links to these pages are in the first 2 positions of the search results. I did not commit any crime and did not inflict physical or material damage to anyone. Nevertheless, the information you control is very troubling me: I work as the leading researcher, teach students and practice phlebology. Currently, this kind of information can play a fatal role in Russia. I request your participation in the removal of the above pages from your site. Help me, please! Thank you and once again I apologize for invading a non-core site. 

Sincerely yours, 

Yuri Gustelev

**Comment by Cavasurg** [comment]
Dear Lenik Terenin!
I greatly apologize for the off-top, but it is very difficult to contact you, although I have repeatedly tried to do it. I have a huge request for you, as the owner of the site http://www.samosud.org. Your resource contains copies of the court decisions: http://www.samosud.org/case_969293378 and http://www.samosud.org/case_849613734, in which I appear as the complainant in the labor dispute. The problem is that while somebody enters my name into the Yandex search box, the links to these pages are in the first 2 positions of the search results. I did not commit any crime and did not inflict physical or material damage to anyone. Nevertheless, the information you control is very troubling me: I work as the leading researcher, teach students and practice phlebology. Currently, this kind of information can play a fatal role in Russia. I request your participation in the removal of the above pages from your site. Help me, please! Thank you and once again I apologize for invading a non-core site. 

Sincerely yours, 

Yuri Gustelev
